pep-0690.rst

```diff
@@ -166,35 +183,205 @@ will be enough to trigger the import of ``spam.py``::
+``sys.modules``. Other than the delayed import, the implementation is
+transparent.
+
+If a module is imported lazily, no entry for it will appear in ``sys.modules``
+at all until it is actually imported on first reference.
+
+If two different modules (``moda`` and ``modb``) both contain a lazy ``import
+foo``, each module's namespace dictionary will have an independent lazy import
+object under the key ``"foo"``, delaying import of the same ``foo`` module. This
+is not a problem. When there is first a reference to, say, ``moda.foo``, the
+module ``foo`` will be imported and placed in ``sys.modules`` as usual, and the
+lazy object under the key ``moda.__dict__["foo"]`` will be replaced by the
+actual module ``foo``. At this point ``modb.__dict__["foo"]`` will remain a lazy
+import object. When ``modb.foo`` is later referenced, it will also try to
+``import foo``. This import will find the module already present in
+``sys.modules``, as is normal for subsequent imports of the same module in
+Python, and at this point will replace the lazy import object at
+``modb.__dict__["foo"]`` with the actual module ``foo``.
+
+There is one case in which a lazy import object can escape one dictionary (but
+only into another dictionary) without being resolved. To preserve the
+performance of ``dict.update()``, it does not check for or resolve lazy import
+objects. However, if the source dict has a ``*_LAZY`` lookup kind set that
+indicates it might contain lazy objects, that lookup kind will be passed on to
+the updated dictionary. This still ensures that the lazy import object can't
+escape into Python code without being resolved.
+
+Other "bulk" dictionary lookup methods (such as ``dict.items()``,
+``dict.values()``, etc) will resolve all lazy import objects in the dictionary.
+Since it is uncommon for any of these to be used on a module namespace
+dictionary, the priority here is simplicity of implementation and minimizing the
+overhead on normal non-lazy dictionaries (just one check to see if the
+dictionary has a ``*_LAZY`` lookup kind).
+
+The eagerness of imports within ``try`` / ``except`` / ``with`` blocks or within
+class or function bodies is handled in the compiler via a new
+``EAGER_IMPORT_NAME`` opcode that always imports eagerly. Top-level imports use
+``IMPORT_NAME``, which may be lazy or eager depending on ``-L`` and/or
+``importlib.enable_lazy_imports()``.
+
+
+Exceptions
+----------
+
+Exceptions that occur during a lazy import bubble up and erase the
+partially-constructed module(s) from ``sys.modules``, just as exceptions during
+normal import do.
+
+Since errors raised during a lazy import will occur later (wherever the imported
+name is first referenced) than they would if the import were eager, it is
+possible that they could be accidentally caught by exception handlers that
+didn't expect the import to be running within their ``try`` block, leading to
+confusion. To reduce the potential for this confusion, errors raised in the
```

Comment:
Is this Exceptions only? If it's also BaseExceptions, then logically LazyImportError should also be a BaseException. I think Guido suggested a separate LazyImportError base class that is a BaseException, similar to BaseExceptionGrooup vs ExceptionGroup. Another design could be to simply not wrap BaseExceptions.

Reply:
Oh shoot, thanks for the reminder, I forgot about that issue. My inclination is to think we don't need to wrap `BaseException`, only `Exception`. If someone is actually catching `KeyboardInterrupt` or `SystemExit`, they will want to catch it within a lazy import as well. These are not "semantic" errors like `TypeError` or `ValueError` where catching it from a location you didn't expect is likely to lead to subtly wrong behavior.

I'll update the PR to clarify.